### ISSUE TITLE
Expand testing 

### ISSUE THREAD

**Issue by MilesCranmer** [JuliaPy/PythonCall.jl#448]
Currently the test suite is very small, most notably the juliacall suite: https://github.com/JuliaPy/PythonCall.jl/blob/main/pytest/test_all.py which is only 19 lines long. I am very worried about reliability issues from this and whether this might result in regressions.

I think expanding the test suite should be a priority, especially on the Python side, before new features worked on.

Here are some suggestions. Feel free to check any that are already done or add more

- [ ] Unittests for features shown in the docs
  - [ ] JuliaCall guide examples
  - [ ] PythonCall guide examples
  - [ ] ALL conversion rules Python -> Julia
  - [ ] ALL conversion rules Julia -> Python
- [ ] Miscellaneous unittests
  - [ ] GC tests (currently there are zero)
  - [ ] Multithreading unittests with signals turned on and GC disabled
  - [ ] Distributed compute unittests
  - [ ] Julia's stdlib tests from Python via juliacall (copy the test suite; see what is broken, and turn off as needed)
  - [ ] Python's stdlib tests from Julia via PythonCall.jl (copy the test suite; see what is broken, and turn off as needed)
  - [ ] Any areas missing due to coverage milestones:
- [ ] Code coverage milestones
  - [ ] Reach 60% code coverage on Python source code
  - [ ] Reach 80% code coverage on Python source code
  - [ ] Set up code coverage for Julia source code (currently not tracked)
  - [ ] Reach 60% code coverage on Julia source code
  - [ ] Reach 80% code coverage on Julia source code
- [ ] Add integration tests
  - [ ] Numpy
  - [ ] PythonPlot.jl
  - [ ] Pandas (https://juliapy.github.io/PythonCall.jl/stable/compat/#Tabular-data-/-Pandas)
  - [ ] SymPy.jl (whenever PythonCall.jl version is made)
  - [ ] Matplotlib interactive support (harder to test, but you would want to test the features explained here – https://juliapy.github.io/PythonCall.jl/stable/compat/#MatPlotLib-/-PyPlot-/-Seaborn)
  - [ ] IPython support
  - [ ] PySR
  - [ ] diffeqpy
  - [ ] PyTorch import issue (basically just verify the solution on https://juliapy.github.io/PythonCall.jl/stable/faq/)

**Comment by cjdoris** [comment]
Absolutely, I've been holding off making another release because I recently did a big refactor and want to get test coverage up so I can trust I didn't break anything.

Indeed the test suite of JuliaCall is short - but then so is JuliaCall itself. My priority has been to get the coverage for PythonCall up, because that covers most of JuliaCall too.

Your list looks really good. I don't think it's especially needed to test the same code twice (from Julia and from Python) so most of the points in the first group should probably done in the PythonCall tests in Julia, leaving the tests in Python to the parts of JuliaCall actually written in Python.

Integration tests sound great.

**Comment by MilesCranmer** [comment]
Sounds good to me!

**Comment by MilesCranmer** [comment]
Sounds good to me!
